### PR TITLE
Created new OEQThumb component

### PR DESCRIPTION
**Pull request by mrblippy**
##### Checklist

- [x] the [contributor license agreement][] is signed
- [x] commit message follows [commit guidelines][]
- [ ] tests are included
- [x] screenshots are included showing significant UI changes

##### Description of change

https://github.com/openequella/openEQUELLA/issues/1306

This commit covers
- oEQThumb component, which will display a real thumbnail for an attachment (if it exists). Otherwise an appropriate placeholder icon is returned.
- Add null check inside SearchHelper.scala. If this is not performed the fileExists method can return incorrect results, when the filename parameter is null oEQ falls back to checking if the item file handle exists (which it always will exists)
- Slight modification to SearchResult component to use the oEQThumb component. We are still always using the first attachment as the display thumbnail. This will be expanded out in future PRs to reflect the SaerchResult's displayOptions.

Tests will be coming in a separate PR, but for now I would really like some feedback on my logic inside the OEQThumb component. I tried my best to reduce the amount of repeated code, but any suggestions further feedback would be welcome

[contributor license agreement]: https://www.apereo.org/node/676
[commit guidelines]: https://chris.beams.io/posts/git-commit/

### CODE REVIEW

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+    <img
+      className={`MuiPaper-elevation1 MuiPaper-rounded ${classes.thumbnail}`}
+      src={links.thumbnail}
+      alt={description}
+    />
+  );
+  /**
+   * @param {string} mimeType - Attachment's Mimetype. eg. image/png application/pdf
+   * @return {ReactElement} Image, video and pdf based mimetypes are thumbnailed by oEQ.
+   * We need to check if a thumbnail has been generated, and return a generic icon if not
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:82]: This should be before the `param` and `return` statements.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+      if (hasGeneratedThumb) {
+        result = oeqProvidedThumb;
+      } else {
+        if (mimeType?.startsWith("image")) {
+          result = <ImageIcon {...generalThumbStyles} />;
+        } else if (mimeType?.startsWith("video")) {
+          result = <VideoIcon {...generalThumbStyles} />;
+        }
+      }
+    }
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:100]: My eye was immediate draw to this block due to what appears to be excessive nesting. As excessive nesting can make code hard to follow/read/understand intent.

Looking at this, I wonder why `if(hasGeneratedThumb)` is in the nesting. It feels as though this whole level can be replaced with an opening statement in this function of:

```typescript
if(hasGeneratedThumb) {
  return oeqProvidedThumb;
}
```

If that is valid, it then looks like this function could be:

```typescript
const handleMimeType = (mimeType?: string): React.ReactElement => {
  if (hasGeneratedThumb) {
    return oeqProvidedThumb;
  }

  let result = <DefaultFileIcon {...generalThumbStyles} />;
  if (mimeType?.startsWith("image")) {
    result = <ImageIcon {...generalThumbStyles} />;
  } else if (mimeType?.startsWith("video")) {
    result = <VideoIcon {...generalThumbStyles} />;
  }

  return result;
}

```

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+      break;
+    }
+    case "custom/youtube": {
+      oeqThumb = oeqProvidedThumb;
+      break;
+    }
+    case "custom/googlebook": {
+      oeqThumb = oeqProvidedThumb;
+      break;
+    }
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:124]: These should all just be a fall-through. E.g.:

```typescript
    case "custom/flickr":
    case "custom/youtube":
    case "custom/googlebook":
      oeqThumb = oeqProvidedThumb;
      break;
```

Also note, no need for the braces on your case statements. (Both here and in the rest of this switch.)

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+  links,
+}: OEQ.Search.Attachment) {
+  const classes = useStyles();
+  const generalThumbStyles: ThumbProps = {
+    className: `MuiPaper-elevation1 MuiPaper-rounded ${classes.thumbnail} ${classes.placeholderThumbnail}`,
+    fontSize: "large",
+  };
+  let oeqThumb: React.ReactElement = (
+    <DefaultFileIcon {...generalThumbStyles} />
+  );
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:70]: Seeing this is a variable, it makes sense to keep down next to where it's changed/used - i.e. down near the switch statement.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx [hunk 5]
```diff
@@ -60,6 +61,8 @@ const useStyles = makeStyles((theme: Theme) => {
     },
     placeholderThumbnail: {
       color: theme.palette.text.secondary,
+      opacity: 0.4,
+      height: "66px",
     },
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:66]: So it's interesting to have this duplicated here to match the `oEQThumb`. Is there something that can be done about this?

I'm kind of wondering if maybe the generation of the 'placeholder' is also part of `OEQThumb`.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx [hunk 6]
```diff
@@ -116,24 +119,16 @@ export default function SearchResult({
...
-        <Subject
-          className={`MuiPaper-elevation1 MuiPaper-rounded ${classes.placeholderThumbnail} ${classes.thumbnail}`}
-          fontSize="large"
-        />
-      );
+      thumb = <OEQThumb {...attachments[0]} />;
     }
-    return result;
+    return thumb;
   };
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:132]: Based on what you have here, we'd really just expect a ternary statement. Such as:

```typescript
const thumbnail = (disableThumb: boolean): React.ReactElement =>
  (attachments.length > 0 && !disableThumb)
    ? <Subject
        className={`MuiPaper-elevation1 MuiPaper-rounded ${classes.placeholderThumbnail} ${classes.thumbnail}`}
        fontSize="large"
      />
    : <OEQThumb {...attachments[0]} />;
```

But as noted above, maybe this functionality is simply part of OEQThumb. :thinking:

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala [hunk 7]
```diff
@@ -224,10 +224,13 @@ object SearchHelper {
...
   def thumbExists(itemKey: ItemIdKey, attachBean: AttachmentBean): Option[Boolean] = {
+
     attachBean match {
       case fileBean: FileAttachmentBean =>
-        val item = LegacyGuice.viewableItemFactory.createNewViewableItem(itemKey)
-        Some(LegacyGuice.fileSystemService.fileExists(item.getFileHandle, fileBean.getThumbnail))
+        val item  = LegacyGuice.viewableItemFactory.createNewViewableItem(itemKey)
+        val thumb = fileBean.getThumbnail
+        if (thumb == null) Some(false)
+        else Some(LegacyGuice.fileSystemService.fileExists(item.getFileHandle, thumb))
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:233]: Can you simply combine these three lines to the more pipeline approach of:

```scala
Option(fileBean.getThumbnail).map { LegacyGuice.fileSystemService.fileExists(item.getFileHandle, _) }
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+    thumbnail: {
+      //if material UI changes such that the MuiPaper-elevation1 MuiPaper-rounded no longer map, add the rules here.
+      marginRight: theme.spacing(2),
+      width: "88px",
+      height: "auto",
+    },
+    placeholderThumbnail: {
+      color: theme.palette.text.secondary,
+      opacity: 0.4,
+      height: "66px",
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:41]: should this height rule be in `thumbnail` above? Or are true oEQ thumbnails sometimes different to that?

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx [hunk 9]
```diff
@@ -116,24 +119,16 @@ export default function SearchResult({
    * Otherwise, return a placeholder icon, to ensure a consistent layout
    */
   const thumbnail = (disableThumb: boolean): React.ReactElement => {
-    let result: any;
+    let thumb: any = (
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:122]: This can be typed as ReactElement.

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx [hunk 10]
```diff
@@ -0,0 +1,131 @@
...
+});
+
+interface ThumbProps {
+  className: string;
+  fontSize: "inherit" | "default" | "small" | "large";
+}
+/**
+ * OEQThumb component
+ * Takes an OEQ.Search.Attachment object as a parameter.
+ * The thumbnail will be served up from oEQ if the attachment has had a thumbnail generated (Image, video, pdf based files, Youtube, Flickr, Goodle books)
```
Comment (SammyIsConfused) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:53]: You have a stray "Goodle" in here :P 
```suggestion
 * The thumbnail will be served up from oEQ if the attachment has had a thumbnail generated (Image, video, PDF based files, YouTube, Flickr, Google books)
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx (hunk 2 above)
Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:100]: That should be valid (and more readable) Thanks for the suggestion. I think maybe I had been looking at it for too long :thinking: :exploding_head:

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx (hunk 4 above)
Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:70]: Yeah, it's sort of hanging out in the middle of nowhere as-is

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx (hunk 5 above)
Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:66]: I did consider putting the placeholder inside the oEQThumb component. But the search placeholder icon is only really relevant to the SearchResult component.

I intended that this oEQThumb component could be used in other areas of the oEQ new ui. Eg. Item summary page, attachment details panel. 
In that case the oEQThumb component would only be used if there were 1 or more attachments, and would never need to produce a placeholder thumbnail.

Not sure if that justification makes sense. So we can have a chat about it if necessary

Comment (edalex-ian) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:66]: Maybe following YAGNI then, how about we put it in there for now, and then when we get to that bit see how things look and rework as required.

Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:66]: Ok, sure thing

File: Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala (hunk 7 above)
Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/scalasrc/com/tle/web/api/search/SearchHelper.scala:233]: Yep, that will do nicely

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx (hunk 10 above)
Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:53]: > You have a stray "Goodle" in here :P

Are you saying goodle is not a real company name :stuck_out_tongue_closed_eyes:

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx (hunk 9 above)
Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:122]: yeah, that's true

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+  InsertDriveFile as DefaultFileIcon,
+  Language as WebIcon,
+} from "@material-ui/icons";
+
+const useStyles = makeStyles((theme: Theme) => {
+  return {
+    thumbnail: {
+      //if material UI changes such that the MuiPaper-elevation1 MuiPaper-rounded no longer map, add the rules here.
+      marginRight: theme.spacing(2),
+      width: "88px",
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:35]: numbers are automatically `px` in material ui

```suggestion
      width: 88,
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx (hunk 1 above)
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:82]:
```suggestion
   * We need to check if a thumbnail has been generated, and return a generic icon if not
   *
   * @param {string} mimeType - Attachment's Mimetype. eg. image/png application/pdf
   * @return {ReactElement} Image, video and pdf based mimetypes are thumbnailed by oEQ.
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx [hunk 12]
```diff
@@ -60,6 +61,8 @@ const useStyles = makeStyles((theme: Theme) => {
     },
     placeholderThumbnail: {
       color: theme.palette.text.secondary,
+      opacity: 0.4,
+      height: "66px",
```
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/SearchResult.tsx:65]:
```suggestion
      height: 66,
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx (hunk 8 above)
Comment (ChristianMurphy) [inline Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:41]:
```suggestion
      height: 66,
```

Comment (mrblippy) [reply Source/Plugins/Core/com.equella.core/js/tsrc/search/components/OEQThumb.tsx:41]: google books are the outlier, which seem to come across in portrait orientation

File: Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx [hunk 13]
```diff
@@ -0,0 +1,144 @@
...
+      height: "66",
+    },
+  };
+});
+
+interface ThumbProps {
+  className: string;
+  fontSize: "inherit" | "default" | "small" | "large";
+}
+interface OEQThumbProps {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx:51]: Nit: can we have a blank line here please?

File: Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx [hunk 14]
```diff
@@ -0,0 +1,144 @@
...
+  /**
+   * On object representing an oEQ attachment. If undefined, a placeholder icon is returned
+   */
+  attachment?: OEQ.Search.Attachment;
+  /**
+   * True indicates that a placeholder icon should be used instead of the 'real' thumbnail
+   */
+  showPlaceholder: boolean;
+}
+/**
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx:61]: Nit: and another above here?

File: Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx [hunk 15]
```diff
@@ -0,0 +1,144 @@
...
+  attachment,
+  showPlaceholder,
+}: OEQThumbProps) {
+  const classes = useStyles();
+  const generalThumbStyles: ThumbProps = {
+    className: `MuiPaper-elevation1 MuiPaper-rounded ${classes.thumbnail} ${classes.placeholderThumbnail}`,
+    fontSize: "large",
+  };
+
+  if (!attachment) {
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx:77]: To make things simpler later, how about here you just go with:

```suggestion
  if (!attachment || showPlaceholder) {
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx [hunk 16]
```diff
@@ -0,0 +1,144 @@
...
+      result = <VideoIcon {...generalThumbStyles} />;
+    }
+    return result;
+  };
+
+  let oeqThumb: React.ReactElement = showPlaceholder ? (
+    <PlaceholderIcon {...generalThumbStyles} />
+  ) : (
+    <DefaultFileIcon {...generalThumbStyles} />
+  );
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx:119]: Couple of things here:

1. With my suggestion above regarding `showPlaceholder`, this no longer needs the ternary; and
2. You've now repeated `<DefaultFileIcon {...generalThumbStyles} />`. You should just move this to a constant up above the `handleMimeType` function so that you can use it in there and here. (i.e. `const defaultThumb = <DefaultFileIcon {...generalThumbStyles} />`)

With that in mind, this block would then just be:

```typescript
let oeqThumb = defaultThumb;
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx [hunk 17]
```diff
@@ -0,0 +1,144 @@
...
+  Language as WebIcon,
+  Subject as PlaceholderIcon,
+} from "@material-ui/icons";
+
+const useStyles = makeStyles((theme: Theme) => {
+  return {
+    thumbnail: {
+      //if material UI changes such that the MuiPaper-elevation1 MuiPaper-rounded no longer map, add the rules here.
+      marginRight: theme.spacing(2),
+      width: "88",
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx:36]:
```suggestion
      width: 88,
```

File: Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+      height: "auto",
+    },
+    placeholderThumbnail: {
+      color: theme.palette.text.secondary,
+      opacity: 0.4,
+      height: "66",
```
Comment (edalex-ian) [inline Source/Plugins/Core/com.equella.core/js/tsrc/components/OEQThumb.tsx:42]:
```suggestion
      height: 66,
```
